### PR TITLE
imgproc: use universal intrinsic in canny

### PR DESCRIPTION
**Pull request by tomoaki0705**
### This pullrequest changes
  * replace SSE2 intrinsic to universal intrinsic
  * add v_abs for universal intrinsic
  * add test of v_abs in test_intrin

I will upload the performance result later.

### CODE REVIEW

File: modules/core/include/opencv2/core/hal/intrin_sse.hpp [hunk 1]
```diff
@@ -739,6 +739,18 @@ inline v_float64x2 v_invsqrt(const v_float64x2& x)
     return v_float64x2(_mm_div_pd(v_1, _mm_sqrt_pd(x.val)));
 }
 
+#define OPENCV_HAL_IMPL_SSE_ABS_INT_FUNC(_Tpvec, func, suffix, subWidth) \
+inline _Tpvec v_abs(const _Tpvec& x) \
+{ return _Tpvec(_mm_##func##_ep##suffix##(x.val, _mm_sub_ep##subWidth##(_mm_setzero_si128(), x.val))); }
```
Comment (alalek) [inline modules/core/include/opencv2/core/hal/intrin_sse.hpp]: `##suffix##` -> `##suffix`
`##subWidth##` -> `##subWidth`

File: modules/imgproc/src/canny.cpp [hunk 2]
```diff
@@ -299,8 +300,8 @@ class parallelCanny : public ParallelLoopBody
 
     void operator()(const Range &boundaries) const
     {
-#if CV_SSE2
-        bool haveSSE2 = checkHardwareSupport(CV_CPU_SSE2);
+#if CV_SIMD128
+        bool haveSIMD = checkHardwareSupport(CV_CPU_SSE2) | checkHardwareSupport(CV_CPU_NEON);
```
Comment (alalek) [inline modules/imgproc/src/canny.cpp]: `|` -> `||` is preferable for boolean values

File: modules/core/include/opencv2/core/hal/intrin_sse.hpp (hunk 1 above)
Comment (tomoaki0705) [inline modules/core/include/opencv2/core/hal/intrin_sse.hpp]: Fixed

**Comment by tomoaki0705** [comment]
Here is the performance test
I used performance test from opencv_perf_imgproc and used the median value from test case ```Img_Aperture_L2_thresholds_canny.canny```

## Comparison on Windows 7 64bit (VS2012 64bit)

test case | before | after
--------- | ------ | -----
("cv/shared/lena.png", 3, false, (50, 100))                                                   | 2.27  |  1.78  | ("cv/shared/lena.png", 3, false, (50, 100))
("cv/shared/lena.png", 3, false, (0, 50))                                                     | 3.60  |  3.32  | ("cv/shared/lena.png", 3, false, (0, 50))
("cv/shared/lena.png", 3, false, (100, 120))                                                  | 1.41  |  1.27  | ("cv/shared/lena.png", 3, false, (100, 120))
("cv/shared/lena.png", 3, true, (50, 100))                                                    | 0.90  |  0.77  | ("cv/shared/lena.png", 3, true, (50, 100))
("cv/shared/lena.png", 3, true, (0, 50))                                                      | 2.01  |  1.59  | ("cv/shared/lena.png", 3, true, (0, 50))
("cv/shared/lena.png", 3, true, (100, 120))                                                   | 0.61  |  0.44  | ("cv/shared/lena.png", 3, true, (100, 120))
("cv/shared/lena.png", 5, false, (50, 100))                                                   | 2.28  |  1.38  | ("cv/shared/lena.png", 5, false, (50, 100))
("cv/shared/lena.png", 5, false, (0, 50))                                                     | 2.30  |  1.37  | ("cv/shared/lena.png", 5, false, (0, 50))
("cv/shared/lena.png", 5, false, (100, 120))                                                  | 2.24  |  1.33  | ("cv/shared/lena.png", 5, false, (100, 120))
("cv/shared/lena.png", 5, true, (50, 100))                                                    | 2.29  |  1.38  | ("cv/shared/lena.png", 5, true, (50, 100))
("cv/shared/lena.png", 5, true, (0, 50))                                                      | 2.31  |  1.38  | ("cv/shared/lena.png", 5, true, (0, 50))
("cv/shared/lena.png", 5, true, (100, 120))                                                   | 2.15  |  1.30  | ("cv/shared/lena.png", 5, true, (100, 120))
("stitching/b1.png", 3, false, (50, 100))                                                     | 2.73  |  2.57  | ("stitching/b1.png", 3, false, (50, 100))
("stitching/b1.png", 3, false, (0, 50))                                                       | 4.30  |  4.04  | ("stitching/b1.png", 3, false, (0, 50))
("stitching/b1.png", 3, false, (100, 120))                                                    | 2.22  |  2.08  | ("stitching/b1.png", 3, false, (100, 120))
("stitching/b1.png", 3, true, (50, 100))                                                      | 1.00  |  0.79  | ("stitching/b1.png", 3, true, (50, 100))
("stitching/b1.png", 3, true, (0, 50))                                                        | 2.63  |  1.78  | ("stitching/b1.png", 3, true, (0, 50))
("stitching/b1.png", 3, true, (100, 120))                                                     | 0.80  |  0.67  | ("stitching/b1.png", 3, true, (100, 120))
("stitching/b1.png", 5, false, (50, 100))                                                     | 3.65  |  2.19  | ("stitching/b1.png", 5, false, (50, 100))
("stitching/b1.png", 5, false, (0, 50))                                                       | 3.78  |  2.36  | ("stitching/b1.png", 5, false, (0, 50))
("stitching/b1.png", 5, false, (100, 120))                                                    | 3.33  |  2.08  | ("stitching/b1.png", 5, false, (100, 120))
("stitching/b1.png", 5, true, (50, 100))                                                      | 3.50  |  2.26  | ("stitching/b1.png", 5, true, (50, 100))
("stitching/b1.png", 5, true, (0, 50))                                                        | 3.82  |  2.25  | ("stitching/b1.png", 5, true, (0, 50))
("stitching/b1.png", 5, true, (100, 120))                                                     | 3.21  |  2.04  | ("stitching/b1.png", 5, true, (100, 120))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (50, 100))  | 5.38  |  4.91  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (50, 100))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (0, 50))    | 7.75  |  7.10  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (0, 50))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (100, 120)) | 4.13  |  3.92  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (100, 120))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (50, 100))   | 2.15  |  1.48  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (50, 100))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (0, 50))     | 3.77  |  2.18  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (0, 50))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (100, 120))  | 1.60  |  1.14  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (100, 120))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (50, 100))  | 4.70  |  3.07  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (50, 100))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (0, 50))    | 4.68  |  3.17  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (0, 50))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (100, 120)) | 4.53  |  2.51  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (100, 120))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (50, 100))   | 4.66  |  3.26  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (50, 100))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (0, 50))     | 4.70  |  3.30  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (0, 50))
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (100, 120))  | 4.51  |  2.53  | ("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (100, 120))


## Comparison on ARM 32bit (Jetson TK1) (GCC 4.8.4 ENABLE_NEON)

test case | before | after
--------- | ------ | -----
("cv/shared/lena.png", 3, false, (50, 100)) |  6.10  |  5.99 
("cv/shared/lena.png", 3, false, (0, 50)) |  10.09  |  10.25 
("cv/shared/lena.png", 3, false, (100, 120)) |  4.97  |  4.60 
("cv/shared/lena.png", 3, true, (50, 100)) |  5.56  |  5.47 
("cv/shared/lena.png", 3, true, (0, 50)) |  9.95  |  10.16 
("cv/shared/lena.png", 3, true, (100, 120)) |  4.61  |  4.28 
("cv/shared/lena.png", 5, false, (50, 100)) |  13.77  |  13.94 
("cv/shared/lena.png", 5, false, (0, 50)) |  13.60  |  13.76 
("cv/shared/lena.png", 5, false, (100, 120)) |  13.30  |  13.48 
("cv/shared/lena.png", 5, true, (50, 100)) |  13.34  |  13.69 
("cv/shared/lena.png", 5, true, (0, 50)) |  13.55  |  13.96 
("cv/shared/lena.png", 5, true, (100, 120)) |  13.06  |  13.39 
("stitching/b1.png", 3, false, (50, 100)) |  9.20  |  8.47 
("stitching/b1.png", 3, false, (0, 50)) |  15.77  |  15.86 
("stitching/b1.png", 3, false, (100, 120)) |  8.15  |  7.29 
("stitching/b1.png", 3, true, (50, 100)) |  8.85  |  8.20 
("stitching/b1.png", 3, true, (0, 50)) |  15.92  |  15.92 
("stitching/b1.png", 3, true, (100, 120)) |  7.83  |  6.98 
("stitching/b1.png", 5, false, (50, 100)) |  23.36  |  23.85 
("stitching/b1.png", 5, false, (0, 50)) |  23.80  |  24.11 
("stitching/b1.png", 5, false, (100, 120)) |  21.60  |  21.83 
("stitching/b1.png", 5, true, (50, 100)) |  21.91  |  22.39 
("stitching/b1.png", 5, true, (0, 50)) |  23.64  |  24.10 
("stitching/b1.png", 5, true, (100, 120)) |  19.44  |  19.46 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (50, 100)) |  14.94  |  14.93 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (0, 50)) |  21.04  |  21.36 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (100, 120)) |  12.64  |  12.16 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (50, 100)) |  13.92  |  13.93 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (0, 50)) |  20.80  |  21.27 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (100, 120)) |  11.96  |  11.42 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (50, 100)) |  27.27  |  27.76 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (0, 50)) |  27.52  |  27.90 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (100, 120)) |  26.83  |  27.27 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (50, 100)) |  27.00  |  27.76 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (0, 50)) |  27.35  |  28.01 
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (100, 120)) |  26.40  |  27.05 


## Comparison on ARM 64bit (ODROID-C2) (GCC 5.4.0)

test case | before | after
--------- | ------ | -----
("cv/shared/lena.png", 3, false, (50, 100)) |  4.29  |  4.96
("cv/shared/lena.png", 3, false, (0, 50)) |  5.07  |  5.42
("cv/shared/lena.png", 3, false, (100, 120)) |  4.78  |  2.98
("cv/shared/lena.png", 3, true, (50, 100)) |  4.24  |  3.58
("cv/shared/lena.png", 3, true, (0, 50)) |  5.66  |  6.91
("cv/shared/lena.png", 3, true, (100, 120)) |  3.44  |  3.78
("cv/shared/lena.png", 5, false, (50, 100)) |  6.86  |  7.39
("cv/shared/lena.png", 5, false, (0, 50)) |  7.22  |  9.14
("cv/shared/lena.png", 5, false, (100, 120)) |  7.15  |  10.00
("cv/shared/lena.png", 5, true, (50, 100)) |  7.15  |  6.66
("cv/shared/lena.png", 5, true, (0, 50)) |  10.07  |  6.68
("cv/shared/lena.png", 5, true, (100, 120)) |  7.89  |  9.93
("stitching/b1.png", 3, false, (50, 100)) |  6.19  |  5.67
("stitching/b1.png", 3, false, (0, 50)) |  8.02  |  8.24
("stitching/b1.png", 3, false, (100, 120)) |  5.94  |  5.13
("stitching/b1.png", 3, true, (50, 100)) |  6.01  |  5.65
("stitching/b1.png", 3, true, (0, 50)) |  8.18  |  8.31
("stitching/b1.png", 3, true, (100, 120)) |  5.65  |  4.48
("stitching/b1.png", 5, false, (50, 100)) |  11.35  |  11.01
("stitching/b1.png", 5, false, (0, 50)) |  11.61  |  11.23
("stitching/b1.png", 5, false, (100, 120)) |  10.94  |  10.15
("stitching/b1.png", 5, true, (50, 100)) |  10.89  |  10.48
("stitching/b1.png", 5, true, (0, 50)) |  11.76  |  11.80
("stitching/b1.png", 5, true, (100, 120)) |  10.05  |  10.32
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (50, 100)) |  9.14  |  10.24
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (0, 50)) |  10.83  |  10.03
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, false, (100, 120)) |  8.26  |  9.29
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (50, 100)) |  8.29  |  9.68
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (0, 50)) |  11.28  |  10.35
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 3, true, (100, 120)) |  8.09  |  8.68
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (50, 100)) |  13.51  |  15.65
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (0, 50)) |  14.07  |  12.99
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, false, (100, 120)) |  13.42  |  13.66
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (50, 100)) |  13.52  |  12.91
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (0, 50)) |  13.78  |  13.10
("cv/detectors_descriptors_evaluation/images_datasets/leuven/img1.png", 5, true, (100, 120)) |  13.21  |  13.36

**Comment by tomoaki0705** [comment]
The point of this PR is to use universal intrinsic, so it won't accelerate the implementation.
Still, the numbers are basically same which is wished to happen.

Best.

**Comment by alalek** [comment]
Thank you! :+1:

**Pull request merged**